### PR TITLE
Add Dynamic Tests for the Artemis Domain Classes

### PR DESCRIPTION
**Pull request by MaisiKoleni**
### Checklist
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context
Currently, some domain class methods are not tested or the tests only print to the console.

We try to add proper tests that adapt to whatever domain classes we have and check some basic invariants and behavioral properties. The main focus are the methods of `java.lang.Object`.

We could also add tests to see if our classes are in fact `Serializable`.

### Description
- Added class path scanning utilities
- Added test class for dynamic test of the domain classes

#### Todo List
- [x] Add exceptions to some rules
- [x] Fix the bugs found by the tests
- [x] Remove `System.out.println` in all the other tests
- [ ] See if https://github.com/gradle/gradle/issues/5975 is an impediment?
       We have many tests with the same name, but different `DynamicContainers`. In the console, the path is not displayed correctly.

### Steps for Testing
1. Review the code, check that you can understand most of what is does.
2. Run the tests on your machine, especially the `ObjectMethodTest` in your IDE.

### Test Coverage

- +12 something according to Codacy
- +300 instructions according to JaCoCo
- **Now vs. Before**
  image

### Screenshots
image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizExercise.java [hunk 1]
```diff
@@ -144,7 +144,7 @@ public String getType() {
     @Override
     @JsonView(QuizView.Before.class)
     public ZonedDateTime getDueDate() {
-        return isPlannedToStart && getReleaseDate() != null ? getReleaseDate().plusSeconds(getDuration()) : super.getDueDate();
+        return isPlannedToStart != null && isPlannedToStart && getReleaseDate() != null ? getReleaseDate().plusSeconds(getDuration()) : super.getDueDate();
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizExercise.java:147]: I would rather suggest to use the getter and change it to lower-case boolean

```
public boolean isIsPlannedToStart() {
    return Boolean.TRUE.equals(isPlannedToStart);
}

```

Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizExercise.java:147]:
```suggestion
        return isIsPlannedToStart() && getReleaseDate() != null ? getReleaseDate().plusSeconds(getDuration()) : super.getDueDate();
```

**Review by krusche — APPROVED** [review]
code looks good to me

**Review by sjagla27 — APPROVED** [review]
Code looks good. Tests passed locally.

**Review by sleiss — APPROVED** [review]
Successfully executed the test locally.

**Review by anditurdiu — APPROVED** [review]
Tests worked fine also on Intellij. Good job 👍

**Review by fde312 — APPROVED** [review]
Tests worked for me on IntelliJ. Good job! 🚀

**Review by filip-gregurevic — APPROVED** [review]

**Review by stefanwaldhauser — APPROVED** [review]
Amazing work!

**Review by balazs-czopf — APPROVED** [review]
All tests worked for me. Nice work 🚀

**Pull request merged**